### PR TITLE
Allow data sharing and definition self-reference through "this" 

### PR DESCRIPTION
**Pull request by kennsippell**
#139

### CODE REVIEW

File: src/nools/lib.js [hunk 1]
```diff
@@ -1,5 +1,34 @@
...
+  var keys = Object.keys(copy);
+  for (var i in keys) {
+    var key = keys[i];
+    if (typeof copy[key] === 'object') {
+      copy[key] = deepCopy(copy[key]);
+    }
+  }
+  return copy;
+}
+
```
Comment (dianabarsan) [inline src/nools/lib.js]: This function doesn't handle arrays well, unfortunately: 
```
console.log(deepCopy({ a: [1, 2, 3] })) 
> { a: { '0': 1, '1': 2, '2': 3 } }
```

File: src/nools/lib.js [hunk 2]
```diff
@@ -1,5 +1,34 @@
+function bindAllFunctionsToContext(obj, context) {
+  var keys = Object.keys(obj);
+  for (var i in keys) {
+    var key = keys[i];
+    switch(typeof obj[key]) {
+      case 'object':
+        bindAllFunctionsToContext(obj[key], context);
+        break;
+      case 'function':
+        obj[key] = obj[key].bind(context);
```
Comment (dianabarsan) [inline src/nools/lib.js]: This means that nobody should ever use arrow functions when configuring. We should make sure people know this?

Comment (kennsippell) [reply src/nools/lib.js]: You can use lambdas and they will be totally supported without issue, just the `this` keyword won't work. It isn't required for most functions. But yes -- it should certainly be documented. I'll make a note of that in the doc changes I'm preparing.

File: src/nools/lib.js (hunk 1 above)
Comment (kennsippell) [reply src/nools/lib.js]: Good catch

Comment (dianabarsan) [reply src/nools/lib.js]: There are still a couple of edge cases where this function fails:
```
console.log(deepCopy([1, 2, 3]));
> { '0': 1, '1': 2, '2': 3 }
console.log(deepCopy({a: [[1], [2], [3]]}));
> { a: [ { '0': 1 }, { '0': 2 }, { '0': 3 } ] }

**Comment by kennsippell** [comment]
Rebased after merge conflict.

**Comment by kennsippell** [comment]
@dianabarsan Just pushed a commit here fixing a weird edge-case bug. I think it's pretty safe, but requesting your re-review.

**Pull request merged**